lib/Conversion/StructuredToMemref/StructuredToMemrefPass.cpp

```diff
@@ -106,14 +200,127 @@ class StructuredToMemrefPass
+  // above:
+  //
+  //  %cast_1 = memref.reinterpret_cast %arg0 to offset [%offset]
+  //  %cast_2 = memref.reinterpret_cast %cast_1 to offset [%offset]
+  //  %cast_3 = memref.reinterpret_cast %cast_2 to offset [%offset]
+  //
+  // The above sequence is equivalent to:
+  //
+  //  %cast_1 = memref.reinterpret_cast %arg0 to offset [%offset]
+  //  %cast_2 = memref.reinterpret_cast %arg0 to offset [%offset]
+  //  %cast_3 = memref.reinterpret_cast %arg0 to offset [%offset]
+  //
+  // In other word, memref.reinterpret_cast ignores the current offset of the
+  // input buffer.
+  //
+  // Therefore, we have to manually track the offset for each addptr by lowering
+  // to the following form:
+  //
+  // %offset_1 = arith.addi %cst_0 %offset
+  // %cast_1 = memref.reinterpret_cast %arg0 to offset [%offset_1]
+  //
+  // %offset_2 = arith.addi %offset_1 %offset
+  // %cast_2 = memref.reinterpret_cast %arg0 to offset [%offset_2]
+  //
+  // %offset_3 = arith.addi %offset_2 %offset
+  // %cast_3 = memref.reinterpret_cast %arg0 to offset [%offset_3]
+  //
+  // Each tt.addptr is lowered to a pair of arith.addi that accumulates the
+  // current offset before using that offset to the reinterpret_cast.
+  LogicalResult convertAddPtrToReinterpretCast() {
+    auto moduleOp = getOperation();
+
+    RewritePatternSet patterns(&getContext());
+
+    auto context = &getContext();
+    OneToNTypeConverter converter;
+    converter.addConversion([](Type type) { return type; });
+
+    // We are doing a 1->2 type conversion here, where a triton pointer type
+    // maps to a pair of {memref, index} type for the the buffer and offset.
+    converter.addConversion(
+        [context](triton::PointerType ptrType, SmallVectorImpl<Type> &types)
+            -> std::optional<LogicalResult> {
+          types = SmallVector<Type>{getMemrefTypeForScalarPtr(ptrType, context),
+                                    IndexType::get(context)};
+          return success();
+        });
+
+    // Hooks to compute the correct materialization, "argument" and "source"
+    // materialization are used when we need to convert a pair of {memref,
+    // index} type back to the original triton pointer type.
+    // These are used when there are ops that still need to use the original
+    // pointer type. For instance, we convert the result of tt.addptr from
+    // tt.ptr type to a pair of {memref, index}, but the original ptr result is
+    // still being used by another tt.load or tt.store.
+    converter.addArgumentMaterialization(buildCastOp);
+    converter.addSourceMaterialization(buildCastOp);
+
+    // Compute the target materialization, given a value with the pointer type,
+    // convert that value to a pair of {memref, index} type.
+    converter.addTargetMaterialization(buildCastAndOffsetOps);
+
+    patterns.add<ScalarAddptrConverter>(converter, context);
+
+    scf::populateSCFStructuralOneToNTypeConversions(converter, patterns);
```

Comment:
does this help update iter-args on loops?

Reply:
yeah this call populates the patterns that help update the iterargs automatically. we don't have any custom code that handles iterargs specifically like in the ptranalysis pass.